### PR TITLE
Proper Selection of Levels; Save/Load of Levels and Level Metadata

### PR DESCRIPTION
**Pull request by jpmartin2**
#### What issue does this fix?
- Initial work on #26 and #33. 
#### What features does this propose?
- Created files for levels 1-15 that now get loaded.
- Selecting a level to play actually uses the right level data (the data from the file).
- Separated level metadata (locked/unlocked, stars earned) from level.
- Added functionality for saving / loading metadata (but these functions aren't being used yet except in tests).
#### Did you add tests to cover the code you propose?
- Added a test for saving / loading level metadata.
- Extended test for level saving / loading.

### CODE REVIEW

**Comment by PeterMitrano** [comment]
Clearly you didn't bother checking what serialVerisionUID is or what it's used for... in this case where we don't want to bother updating them when we change API, it is better to leave it unassigned then to assign them all the same value (the calculated value will be more useful). [source](http://stackoverflow.com/questions/285793/what-is-a-serialversionuid-and-why-should-i-use-it#answer-285827)

File: src/main/java/aeneas/models/Level.java [hunk 1]
```diff
@@ -27,25 +52,6 @@ public Level(Bullpen bullpen, boolean prebuilt) {
...
     this(bullpen, true);
   }
-  
-  @Override
-  public int hashCode() {
-    return levelNumber;
-  }
-  
-  @Override
-  public boolean equals(Object o) {
```
Comment (PeterMitrano) [inline src/main/java/aeneas/models/Level.java:45]: why was this removed?

**Comment by jpmartin2** [comment]
I have actually read about what serialVersionUID is for.

From the Oracle Docs:

> ... it is strongly recommended that all serializable classes explicitly declare serialVersionUID values, since the default serialVersionUID computation is highly sensitive to class details that may vary depending on compiler implementations ...

The main reason I set them all was because during the process I had added a few getters - this made the version numbers change and the objects become incompatible. I think it is unlikely that the data will change, so I thought it would be preferable to have the versions not change if we decide to add a public helper method later.

File: src/main/java/aeneas/models/Level.java (hunk 1 above)
Comment (jpmartin2) [inline src/main/java/aeneas/models/Level.java:45]: Because after discussion with James, I had moved the storage of metadata to be a hash map of Integer->Metadata, instead of Level->Metadata, and levels are stored in a similar map - since the Level is not a key, we don't need hashCode. 

I'm not sure if the current structure is the best, so I'm open to further ideas on that front.

File: src/main/java/aeneas/models/Model.java [hunk 2]
```diff
@@ -24,40 +31,25 @@
...
 
-    //add custom level for example
-    for (int i=0;i<20;i++){
-      Bullpen b = new Bullpen(new ArrayList<>());
-      Level l = new PuzzleLevel(b, false);
-      l.prebuilt = false;
-      l.levelNumber = i+16;
-      levels.add(l);
-    }
+    System.out.println(index.getLevel(2).getPieces());
```
Comment (PeterMitrano) [inline src/main/java/aeneas/models/Model.java]: plz remove all the printlns

Comment (jpmartin2) [inline src/main/java/aeneas/models/Model.java]: Ooops. Thought I had gotten those.

**Comment by PeterMitrano** [comment]
Ok, so I would argue we do want it to be invalid when we change stuff. I think it is likely the data will change, and more importantly I don't want to worry about updating that number when it does. I haven't had to test with it but I don't like the idea of keeping track of it ourselves when java will give us a suitable version on its own.

File: src/main/java/aeneas/controllers/SelectLevelController.java [hunk 3]
```diff
@@ -13,20 +14,21 @@
...
     this.view = view;
   }
 
   @Override
   public void handle(MouseEvent event) {
-    System.out.println("enter level " + levelModel.levelNumber);
-    if (!levelModel.isLocked()) {
-      view.switchToPlayLevelView();
-      resetLevel();
+    System.out.println("enter level " + levelModel.getLevelNumber());
```
Comment (PeterMitrano) [inline src/main/java/aeneas/controllers/SelectLevelController.java]: still printlns

File: src/main/java/aeneas/views/MainView.java [hunk 4]
```diff
@@ -123,7 +126,10 @@ public void switchToBuildLevelView() {
 
   }
 
-  public void switchToPlayLevelView() {
+  public void switchToPlayLevelView(Level level) {
+    System.out.println("LEVEL: "+level.getLevelNumber());
```
Comment (PeterMitrano) [inline src/main/java/aeneas/views/MainView.java]: more printlns

**Comment by jpmartin2** [comment]
I can understand that, though one part from the docs that stood out was

> depending on compiler implementations

This is probably not going to affect us but we are using multiple implementations (oracle java vs openjdk) and while it may be unlikely that this will cause issues, it is possible and could suck a lot if it did.

In any case, I don't think it'll be a big deal to bump the numbers when things change.

File: src/main/java/aeneas/controllers/SelectLevelController.java (hunk 3 above)
Comment (jpmartin2) [inline src/main/java/aeneas/controllers/SelectLevelController.java]: That's one you added a while back :P

Should be fixed now though

**Comment by PeterMitrano** [comment]
Ok, I'm ok leaving it as is.

**Comment by jkuszmaul** [comment]
I'm fine with it overall, but is there any simple way that we could create a test or something that would fail if the serialVersionUID was not properly updated? I'm thinking maybe have a test file that is generated off of the master branch and the current code has to either be able to read it or fail due to a mismatched serialVersionUID (and the test itself would fail if the serial versions matched but the class didn't load properly).
Or maybe a commit hook that any time you modify one of the serialVersion files it asks if you remembered to update the UID as appropriate.

**Comment by jpmartin2** [comment]
I'd like to write a test that makes sure the numbers get bumped when necessary, but I haven't found a clean way of doing that yet.

The problem is that the deserialization won't fail even if there are missing/extra fields, so if a new field gets added - lets pretend a level needs a second reference to a different bullpen, that has some default non-null value - the deserialized level will have that field as null.

**Comment by jpmartin2** [comment]
Ok, so for now I've removed the serialVersionUIDs (at the request of Peter). If we decide to, we can add them back in later.

**Comment by PeterMitrano** [comment]
Why are these tests failing? also update this plz

**Comment by jpmartin2** [comment]
Forgot to update the level files

**Comment by jpmartin2** [comment]
Hmmm. those tests pass locally.....

Just not when running with gradle? I'm not sure why

**Comment by PeterMitrano** [comment]
Maybe it actually is the serialization stuff. Try adding the numbers back
and see if the tests pass

**Comment by jpmartin2** [comment]
After merging with master, everything seems fine

**Comment by PeterMitrano** [comment]
@jkuszmaul review and merge plz thx
